### DISCUSSION TITLE
How can I do with Warning: Prop `disabled` did not match. Server: "null" Client: "true"?

### DISCUSSION THREAD

**Discussion by ahadrt** [vercel/next.js#21999] [Help]
I always  get Warning: Prop `disabled` did not match. Server: "null" Client: "true" message and when I  check the code every thing is in order. I even checked out the server side value of prob using code ` if (!process.browser){console.log(5566,"is: ",ready);}` and realized that the value of ready is boolean. 

#### آHow can I deal with this persistent warning?

sample code is like this:
``` TS
  <Fab disabled={!ready } onClick={handleNew} color="primary" size="small">
            {showSelect && <Add></Add>}
            {!showSelect && <CancelOutlined></CancelOutlined>}
          </Fab>
```

**Comment by jrolfs** [comment]
@ahadrt are you seeing this in all browsers? I recently ran into something very similar issue, however, I'm not using Next.js and I only saw the warning in Firefox. I banged my head on this for some time and have come to the conclusion that it's likely a bug that affects React in combination with Firefox. I also would only see the warning when the browser cache was hot (⌘ + Shift + R would prevent the warning from showing).

What does `Fab` ultimately render? In my case, I was rendering a `button` and the issue persisted even when I rendered a `<button />` directly.

What browser and version of React was affected for you?

React @ 17.0.2
Firefox @ 88.0b6 (64-bit)

**Reply by ahadrt** [reply]
@jrolfs I saw it in both chrome and fire fox. I use **material-ui**, so it thought  to be caused by that. I had to change my code because  l couldn't  find any solution for a long run. however, regarding your comment there seems to be a bug in the browser or a **react-browser compatibility** issue. 
 I use react @ 17.0.1, Firefox @ 84.0.1, chrome: @ 89 both 64 , mac os 11.2.3 , and i came across the same warning  in windows as well
I will share here if i find any solution. 
thank you

**Reply by GrantGryczan** [reply]
Whoa, same exact issue, @jrolfs. I'm having the issue you're describing in Firefox but not Chromium. It only happens sometimes though. I can't figure out the cause, and if you're reading this, I probably still haven't found a good solution.

**Reply by justadev999** [reply]
Definitely one of the most helpful answer found! Thank you man!!

**Comment by jrolfs** [comment]
Yeah, I also had to change my code to work around this issue. Based on the snippet you provided above, I suspect we were attempting to implement similar behavior. Does `ready` come from `useState`? I was attempting to disable form submission until the client JS has been hydrated... so I was rendering the `<button />` on the client _and_ server with disabled set to `true` via state and then updating said state to `false` via `useEffect`. Weirdly the hydration mismatch error only occurred when I attempt to update the state via the state setter on the client... no matter when I did so (this is what really felt like a React issue to me). I even tried delaying the `setDisabled` call with a `setTimeout` of a couple of seconds.

**Comment by kairos666** [comment]
Hello there!

I've got the exact same issue. Seems indeed to have a compiler race condition or something.
if I ensure on serverside to provide a primitive instead of a variable pointer it works.

I could shut down the warning with this workaround:
`<Fab disabled={ process.browser ? !ready : true } onClick={handleNew} color="primary" size="small">`

**Reply by deniyaldanidan** [reply]
#### Thank you, Your solution solved my problem... I've used

```js
<button className={styles.secondary} onClick={addToCartFN} disabled={typeof window !== "undefined" ? (inCart + chosenQty > maxProductLimit) : true}>
```

**Comment by glzcc** [comment]
I miss same problem
I use firefox, other normal

**Comment by iamlana** [comment]
I have same issue and it seems to be something with autocomplete in Firefox, which "remembers" the steps, you already filled up in the form. Unfortunately autocomplete=off doesn't help as well. 🤷‍♀️

**Comment by icyJoseph** [comment]
Hi,

This issue could happen if you do something like this:


```jsx
const [ready, setReady] = useState(typeof window !== 'undefined');

return <div className={ready ? 'ready-css' : 'not-ready-css'></div>
```

1. During SSR, type of window is `'undefined'`, so `ready` is false.
2. The SSR process outputs a `div` with a class name `not-ready-css`.
3. The server sends this to the client.
4. The client renders its own copy, and now `typeof window` is an object, so `ready` is true
5. The client sees `<div className="ready-css"></div>`
6. React compares the server side rendered node and the client rendered node, their className missmatches
7. A warning is printed.

I know this explanation might too simplistic, but somehow, somewhere in your application, this is happening. May not be with window, but something else. For example: https://github.com/vercel/next.js/discussions/34982.

**Reply by jrolfs** [reply]
Yeah, I'm definitely aware of how this sort of thing can cause hydration mismatches, but what I was experiencing was much more complex (as far as I could tell). I was initializing the state with `useState(false)` to avoid that, and updating said state with `useEffect` (which only runs on the client).

I just dug up my workaround and it looks like I ended up just disabling form submission be setting `type="button"` on the button instead of using the `disabled` attribute. I have no idea why that works around the issue as I'm still using the same technique I explained above.

**Reply by Enkratia** [reply]
@jrolfs hello, did you find solution for this?

**Reply by jrolfs** [reply]
Only the workaround I shared above.

**Reply by Enkratia** [reply]
@icyJoseph hello,  this problem is happening when `streaming` is used. No `new Date()` or `window` or `Array.sort()`, 
No `streaming` - no error. 

If you have 2-4 minutes of free time you can check it even with default next.js project, just add:
1. one `fetch` request
2. `<Suspense>` for server component (to make streaming possible)
3. client component (inside of server component) with state:  `[isDIsabled, setIsDisabled] = React.useState(false)`
4. `<button disabled={state}></button>` (inside of client component)
5. `React.useEffect` with `setIsDisabled(true)`

**Reply by icyJoseph** [reply]
Could you set up a Stackblitz reproduction example? Navigate to next.new on your browser, and then just post back the link to it

**Reply by Enkratia** [reply]
@icyJoseph thank you for response. I find out this problem happens only when `redux toolkit` is used, but works normally with native `react context`. I will redirect my question  to `redux-toolkit`.

**Comment by kkntl** [comment]
I managed to reliably reproduce this problem on my machine, when using [next-pwa](https://www.npmjs.com/package/next-pwa).
I only happens in Firefox (v113.0.2), and only when using a service worker.

package.json:
```
{
    "scripts": {
        "dev": "next dev",
        "build": "next build",
        "start": "next start",
        "lint": "next lint"
    },
    "dependencies": {
        "next": "13.4.4",
        "next-pwa": "5.6.0",
        "react": "18.2.0",
        "react-dom": "18.2.0"
    }
}
```
next.config.js:
```
const createWithPwa = require('next-pwa');
const runtimeCaching = require('next-pwa/cache')

const withPWA = createWithPwa({
    dest: 'public',
    runtimeCaching,
});

module.exports = withPWA({});
```
/pages/Test.jsx:
```
export default function(){
    const [ disabled, setDisabled ] = useState( true );

    useEffect(() => {
        setDisabled( false );
    }, []);

    return <button disabled={ disabled } >
        button
    </button>;
};
```

**Now reproduce whit these steps:**
- run `next dev`
- (Just to be on the safe side delete any old service workers and the `.next` folder)
- open `localhost:3000/Test`
- reload `localhost:3000/Test` again, by putting the cursor into the address bar and press the RETURN key (don't do CTRL-F5, that's not enough)
- reload the page again with F5
- --> now the hydration warning should come up in the console

**The console message:**
```
Warning: Prop `disabled` did not match. Server: "null" Client: "true"
button
__WEBPACK_DEFAULT_EXPORT__@webpack-internal:///./pages/Test.jsx:12:84
App@webpack-internal:///./node_modules/next/dist/pages/_app.js:24:1
PathnameContextProviderAdapter@webpack-internal:///./node_modules/next/dist/shared/lib/router/adapters.js:74:44
ErrorBoundary@webpack-internal:///./node_modules/next/dist/compiled/@next/react-dev-overlay/dist/client.js:305:63
ReactDevOverlay@webpack-internal:///./node_modules/next/dist/compiled/@next/react-dev-overlay/dist/client.js:854:908
Container@webpack-internal:///./node_modules/next/dist/client/index.js:77:1
AppContainer@webpack-internal:///./node_modules/next/dist/client/index.js:181:25
Root@webpack-internal:///./node_modules/next/dist/client/index.js:359:37
```

**Comment by kkntl** [comment]
I found some leads, but none of them alone seem to be the actual problem.

### button with `autoComplete`

[MDN says](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attributes):

> `autocomplete`
> This attribute on a [button](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button) is nonstandard and Firefox-specific. Unlike other browsers, [Firefox persists the dynamic disabled state](https://stackoverflow.com/questions/5985839/bug-with-firefox-disabled-attribute-of-input-not-resetting-when-refreshing) of a [button](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button) across page loads. Setting autocomplete="off" on the button disables this feature; see [Firefox bug 654072](https://bugzil.la/654072).
>
> `disabled`
> Firefox, unlike other browsers, [persists the dynamic disabled state](https://stackoverflow.com/questions/5985839/bug-with-firefox-disabled-attribute-of-input-not-resetting-when-refreshing) of a [button](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button) across page loads. To control this feature, use the [autocomplete](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#autocomplete) attribute.

Indeed, if I add `autoComplete="off"` to the button, the warning doesn't come anymore:
```
<button disabled={ disabled } autoComplete="off">button</button>
```
But `autoComplete` is non-standard, and validators and type checkers might not allow it.

### `null`

Anyway, the warning complains about `null` on the server versus `true` on the client, and I would expect that the `disabled` property in the DOM would always be `true` or `false`, but never `null`.

The HTML code is `<button disabled="">`, which apparently correctly evaluates to `button.disabled: true` in the DOM, also in Firefox, but I suspect that at some point in the service worker the empty string `""` gets evaluated to `null`, either when written to or when read from the cache ?

### thoughts

On the other hand, I wonder why `autoComplete="off"` helps, but the also not using a service worker helps.  
Also I find it weird that without the `setDisabled( false )` inside the `useEffect` the warning doesn't come as well.

**Comment by JoosepAlviste** [comment]
I ran into the same issue, trying to do the exact same thing as @jrolfs (using `useState` and `useEffect` to initially default the disabled state to `true` and to change it to `false` right away after hydration) where it was only broken in Firefox. Adding `autoComplete="off"` to `<button>` and `<input>` elements fixed the issue for me, but I still wanted to keep the autocomplete, so I simply made that also conditional on the `disabled` state:


```tsx
const { isSSR } = useContext(SSRContext)

return (
  <>
    <button disabled={isSSR} autoComplete="off">...</button>
    <input disabled={isSSR} autoComplete={isSSR ? 'off' : undefined} />
  </>
)
```

Where the `SSRContext` provider sets up the `useState` and `useEffect`:

```ts
const [isSSR, setIsSSR] = useState(true)
useEffect(() => {
  setIsSSR(false)
}, [])
```

Now there are no more warnings in the console for me and autocompletion seems to work, thanks for the notes everyone 🎉

**Comment by Heracraft** [comment]
Ran into the same issue today. Its been days actually. Today's the one day I considered it to be a browser related bug. I'm on Zen browser 1.19.3b (Firefox 148.0.2), next 16.2.1 and react 19.2.4.

I don't need autoComplete so turning it off was enough.

**Comment by sefin23** [comment]
---

## Cause 1: Firefox's `autocomplete` persistence bug (most common)

This is a long-standing **Firefox-specific behavior** documented in [Firefox bug 654072](https://bugzil.la/654072). Firefox persists the `disabled` state of `<button>` and `<input>` elements across page loads. When a service worker is involved (next-pwa or any PWA setup), this becomes even more reliably triggered.

The sequence:
1. Server renders `<button disabled={false}>` → outputs `<button>` (no `disabled` attribute → `null` from React's perspective)
2. Firefox restores the previously remembered `disabled` state → button is `disabled=true` in the DOM before React hydrates
3. React compares server HTML (`null`) vs. DOM (`true`) → mismatch warning

**Fix A — Add `autoComplete="off"` to the button** (confirmed by @kkntl, @JoosepAlviste, @Heracraft):
```tsx
<button disabled={isLoading} autoComplete="off">
  Submit
</button>
```
[MDN documents this](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#autocomplete) specifically for this Firefox behavior. It's non-standard for buttons but works.

**Fix B — Conditional `autoComplete`** (from @JoosepAlviste, keeps autocomplete for non-disabled state):
```tsx
const [isSSR, setIsSSR] = useState(true)
useEffect(() => { setIsSSR(false) }, [])

<button
  disabled={isSSR}
  autoComplete={isSSR ? 'off' : undefined}
>
  Submit
</button>
```

---

## Cause 2: State initialized from browser-only values (all browsers)

As explained by @icyJoseph (maintainer), the warning also happens when `useState` is initialized with a value that differs between server and client:

```tsx
// ❌ Causes mismatch
const [ready, setReady] = useState(typeof window !== 'undefined')
// Server: false, Client: true → mismatch
```

**Fix — Initialize consistently, update in `useEffect`:**
```tsx
// ✅ Both server and client start with the same value
const [isClient, setIsClient] = useState(false)
useEffect(() => { setIsClient(true) }, [])

<button disabled={!isClient}>Submit</button>
```

Alternatively, @kairos666's approach of forcing a primitive on the server side:
```tsx
<Fab disabled={process.browser ? !ready : true}>
```
This ensures the server always renders `disabled={true}` consistently.

---

## Cause 3: Redux Toolkit + React Streaming (App Router)

@Enkratia found this triggers specifically when **Redux Toolkit is used with streaming** (Suspense/App Router). The RTK store state differs between server and client during streaming hydration. Switching to native React Context resolves it.

---

## Quick decision guide

| Symptom | Likely Cause | Fix |
|---------|-------------|-----|
| Firefox only, not Chrome | Firefox `autocomplete` persistence | `autoComplete="off"` on button |
| Happens in all browsers | State from browser-only value | Initialize state identically on both sides, update via `useEffect` |
| Started after adding PWA/service worker | Firefox + SW caching | `autoComplete="off"` on button |
| Started after adding Redux Toolkit + streaming | RTK hydration with RSC | Switch to React Context |

**Comment by icyJoseph** [comment]
Let's sum up what's going on here... 

In Firefox, exclusively as per MDN claim:

> Firefox will, unlike other browsers, persist the dynamic disabled state of a [<button>](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/button) across page loads. Use the [autocomplete](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Attributes/autocomplete) attribute to control this feature. [source](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Attributes/disabled)

An initially disabled state, turned to enabled, is persisted by Firefox on page reload.

Next.js always sends the right initial frame, but Firefox corrects disabled states to whatever it has stored in memory. It attempts to persist user progression through forms and such.

Note that this doesn't seem to happen if the button becomes disabled at runtime. After a reload, the button respects the initial HTML frame (enabled).

If you do make the button disabled again, in code, then reloading won't hit the hydration issue. 

Looks like the only sensible way out of this, is the autocomplete attribute workaround. 

Reading through the bug report, https://bugzilla.mozilla.org/show_bug.cgi?id=654072, it does sound like Mozilla devs claim this is a feature. Note that this dates 14 years ago, and a lot of the comments are not related to React apps at all. 

- https://github.com/vercel/next.js/discussions/21999#discussioncomment-6072413 -> This is the real root cause analysis

The other message I marked spam, is just collecting a bunch of other plausible inputs. @kkntl you did find the issue. What makes it tricky is that:

- if the cycle is, `disabled -> enabled -> disabled`, then you won't see any issue
- the cycle that trigger this is `disabled -> enabled` - happens for input tags too

Why is React/Next.js reporting `null` when the value is false, I am not sure but the root cause is FF's behavior.

Throughout the FF issue report, the `autocomplete` (`autoComplete`) workaround is suggested, and does work, because then FF stops persisting user progression.
